kubevirt-vmware/pkg/controller/ovirtprovider/ovirtprovider_controller.go

```diff
@@ -0,0 +1,280 @@
+	reqLogger.Info("Connection secret retrieved.")
+
+	if len(instance.Spec.Vms) == 0 { // list of oVirt VMs is requested to be retrieved
+		err = r.readVmsList(request, connectionSecret)
+		if err != nil {
+			reqLogger.Error(err, "Failed to read list of oVirt VMs.")
+			return reconcile.Result{}, err // request will be re-queued
+		}
+
+		return reconcile.Result{}, nil
+	}
+
+	// secret is present, list of VMs is available, let's check for  details to be retrieved
+	var lastError error = nil
+	for _, vm := range instance.Spec.Vms { // sequential read is probably good enough, just a single VM or a few of them are expected to be retrieved this way
+		if vm.DetailRequest {
+			err = r.readVMDetail(request, connectionSecret, &vm)
+			if err != nil {
+				reqLogger.Error(err, fmt.Sprintf("Failed to read '%s' vm details.", vm.Name))
+				lastError = err
+			}
+		}
+	}
+
+	return reconcile.Result{}, lastError
+}
+
+func (r *ReconcileOVirtProvider) fetchSecret(provider *v2vv1alpha1.OVirtProvider) (*corev1.Secret, error) {
+	secret := &corev1.Secret{}
+	err := r.client.Get(context.TODO(), types.NamespacedName{Name: provider.Spec.Connection, Namespace: provider.Namespace}, secret)
+	return secret, err
+}
+
+func getClient(ctx context.Context, secret *corev1.Secret) (*Client, error) {
+	return NewClient(ctx, string(secret.Data["apiUrl"]), string(secret.Data["username"]), string(secret.Data["password"]), string(secret.Data["ca"]))
+}
+
+// read whole list at once
+func (r *ReconcileOVirtProvider) readVmsList(request reconcile.Request, connectionSecret *corev1.Secret) error {
+	log.Info("readVmsList()")
+
+	r.updateStatusPhase(request, v2vv1alpha1.PhaseConnecting)
+	client, err := getClient(context.Background(), connectionSecret)
+	if err != nil {
+		r.updateStatusPhase(request, v2vv1alpha1.PhaseConnectionFailed)
+		return err
+	}
+	defer client.Close()
+
+	r.updateStatusPhase(request, v2vv1alpha1.PhaseLoadingVmsList)
+	vms, err := client.GetVMs()
+	if err != nil {
+		r.updateStatusPhase(request, v2vv1alpha1.PhaseLoadingVmsListFailed)
+		return err
+	}
+
+	err = r.updateVmsList(request, vms, utils.MaxRetryCount)
+	if err != nil {
+		r.updateStatusPhase(request, v2vv1alpha1.PhaseLoadingVmsListFailed)
+		return err
+	}
+
+	r.updateStatusPhase(request, v2vv1alpha1.PhaseConnectionSuccessful)
+	return nil
+}
+
+func (r *ReconcileOVirtProvider) updateVmsList(request reconcile.Request, vms map[string]string, retryCount int) error {
+	instance := &v2vv1alpha1.OVirtProvider{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		log.Error(err, fmt.Sprintf("Failed to get provider object to update list of VMs, intended to write: '%s'", vms))
+		if retryCount > 0 {
+			utils.SleepBeforeRetry()
+			return r.updateVmsList(request, vms, retryCount-1)
+		}
+		return err
+	}
+
+	instance.Spec.Vms = make([]v2vv1alpha1.OVirtVM, len(vms))
+	for vmID, vmName := range vms {
+		instance.Spec.Vms = append(instance.Spec.Vms, v2vv1alpha1.OVirtVM{
```

Comment:
Much better, but we still need at least cluster name and id. Otherwise the user would not know which VM belongs to which cluster when name collision happens.

Shouldn't we also add datacenter for beter UX?

Reply:
vm id is there but if we want to show cluster name I will add.